supporting namespaces

shireenf-ibm opened the pull request:
Hi,
this is a first commit supporting namespaces and updating some existing examples to be namespaced

I would like to get first review before adding more tests

zivnevo commented on src/baseline_rule.py:
```diff
@@ -89,33 +91,53 @@ def matches_target(self, labels):
+                return {'namespaceSelector': {}}
             return {}
         if isinstance(selectors, IpSelector):
+            if ns_selector:
+                print('Error: namespaceSelector might not be a CIDR')
+                return None
```
This should better be checked in the constructor

zivnevo commented on src/baseline_rule.py:
```diff
@@ -89,33 +91,53 @@ def matches_target(self, labels):
+    def _get_namespace_name_from_selector(self, sel):
+        k8s_ns_label = 'kubernetes.io/metadata.name'
+        selector_dict = self.selectors_as_netpol_peer(sel, True)
+        return selector_dict['namespaceSelector']['matchLabels'].get(k8s_ns_label) or None
```
You might get a dictionary with just `matchExpression` and not `matchLabels`

shireenf-ibm replied:
a question:
lets assume the policy type is Ingress
Should I still build a **k8s networkpolicy** if the target's namespaceSelector contains other labels/expressions than the name of namespace? (i.e. the returned `selector_dict` Is not in the form of matchLabels with only one key: value and `key==kubernetes.io/metadata.name`)
example:

>  target_ns =kubernetes.io/metadata.name=<ns_name>, user = alice 

with current case the metadata will include namespace: <ns_name>
but the other label will be lost!! (Is there a way to add it to the metadata?)

zivnevo replied:
Yes, you are right. It's more complicated than I thought.
@adisos , what do you think? should we give up synthesizing baseline rules into K8s NetworkPolicies and only support synthesizing into Calico's GlobalNetworkPolicies? I believe handling namespaces will be much easier this way.

adisos replied:
Yes, it does sound simpler this way.

zivnevo replied:
@shireenf-ibm - Let's remove `BaselineRule.to_netpol()` and all the methods solely used by it.
We'll leave `BaselineRule.to_global_netpol_calico()`. With Calico you can simply add `namespaceSelector` to both `source` and `destination` (if required).

shireenf-ibm replied:
Hi,
In my last commit I've removed the `BaselineRule.to_netpol()` method.
but kept following methods, because I found them references in other repository in np-guard:
1. `selectors_as_netpol_peer` (kept the 'master' branch version)
2. `get_port_array`

other methods used by `to_netpol()` were removed, including two methods (`sources_as_netpol_peer and targets_as_netpol_peer`) , which were not used at all (not even before)

zivnevo reviewed (changes requested):
Note that namespace selectors should be general selectors, not just in the form `kubernetes.io/metadata.name=<ns_name>`. This means that when converting a BaselineRule to k8s NetworkPolicy you may or may not be able to detect its namespace.

- [ ]  In the examples. also have an example where the namespace selector uses a label different than `kubernetes.io/metadata.name`
- [ ] Also have an example with a namespace selector with a syntax different than `label=value`.
- [ ]  In `src/baseline_rule.py`, in `BaselineRule.__init__()` check that the namespace selectors are not IPBlocks (throw an exception otherwise)
- [ ]  Also in this constructor, check that if the source/target selector is IPBlock then the corresponding namespace selector is null (throw an exception otherwise)
- [ ]  The method `BaselineRule.matches_connection()` has to be modified so it now takes source and destination namespace-labels as well (with a default `None` value for both).  The method should then check whether the rule's namespace selectors match these namespace labels.

shireenf-ibm commented:
Hi @zivnevo 
In addition to fixes regarding namespaceSelector and removing the BaselineRule.to_netpol(). I made a fix (I thought it is correct to do that) and need your confirmation/ dis about it.

1. I noticed that the "action" printed to  the GlobalNetworkPolicy in `BaselineRule.to_global_netpol_calico(self)` was always Allow even if the rule's action was deny. So I fixed that and made relevant changes in the `tests/testBaselineRuleToGlobalNetPol.py` file.

(an example with the old behavior that printed Allow instead of Deny is `examples/restrict_access_to_payment.yaml`.

2. Also, 'protocol' field was typed only for one direction (the first policy created), so I added it to the other direction..Is that correct?

thanks

adisos commented:
> Hi @zivnevo In addition to fixes regarding namespaceSelector and removing the BaselineRule.to_netpol(). I made a fix (I thought it is correct to do that) and need your confirmation/ dis about it.
> 
>     1. I noticed that the "action" printed to  the GlobalNetworkPolicy in `BaselineRule.to_global_netpol_calico(self)` was always Allow even if the rule's action was deny. So I fixed that and made relevant changes in the `tests/testBaselineRuleToGlobalNetPol.py` file.
> 
> 
> (an example with the old behavior that printed Allow instead of Deny is `examples/restrict_access_to_payment.yaml`.
> 
>     2. Also, 'protocol' field was typed only for one direction (the first policy created), so I added it to the other direction..Is that correct?
> 
> 
> thanks

Hi @shireenf-ibm,
1. The action of 'allow' was not a mistake, the baseline rule action allow/deny is referred by the generated query permit/forbid correspondingly.
2. It should be sufficient to have the protocol in just one direction, since the allowed connections are the intersection of connections from both directions.

shireenf-ibm commented:
> > Hi @zivnevo In addition to fixes regarding namespaceSelector and removing the BaselineRule.to_netpol(). I made a fix (I thought it is correct to do that) and need your confirmation/ dis about it.
> > ```
> > 1. I noticed that the "action" printed to  the GlobalNetworkPolicy in `BaselineRule.to_global_netpol_calico(self)` was always Allow even if the rule's action was deny. So I fixed that and made relevant changes in the `tests/testBaselineRuleToGlobalNetPol.py` file.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > (an example with the old behavior that printed Allow instead of Deny is `examples/restrict_access_to_payment.yaml`.
> > ```
> > 2. Also, 'protocol' field was typed only for one direction (the first policy created), so I added it to the other direction..Is that correct?
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > thanks
> 
> Hi @shireenf-ibm,
> 
>     1. The action of 'allow' was not a mistake, the baseline rule action allow/deny is referred by the generated query permit/forbid correspondingly.
> 
>     2. It should be sufficient to have the protocol in just one direction, since the allowed connections are the intersection of connections from both directions.

Thanks, reverting changes

adisos commented on src/baseline_rule.py:
```diff
@@ -101,35 +142,25 @@ def selectors_as_netpol_peer(selectors):
 
     @staticmethod
-    def _selectors_as_netpol_peer_calico(selectors, limit_to_all_expr=True):
+    def _selectors_as_netpol_peer_calico(selectors, limit_to_all_expr=True, ns_selector=False):
```
update docstring

adisos reviewed (commented):
Hi @shireenf-ibm , in the log of the unit tests testBaselineRuleToGlobalNetPol.py it seems there are some errors fetching the new baseline rules.

shireenf-ibm commented:
> Hi @shireenf-ibm , in the log of the unit tests testBaselineRuleToGlobalNetPol.py it seems there are some errors fetching the new baseline rules.

True, because I've added some examples and it tries to read them from the master branch so it doesn't find them yet. (I could test them locally, with replacing the path to my local fs)

shireenf-ibm commented:
Since I saw this tests/testBaselineRules.py file now, 
I want to add a new good test to it -> to read a content from raw.github url
(A thing that I supported in a previous issue, but since I've removed the tests/testBaselineRuleToNetPol.py file we have no example of reading from raw.github urls anymore)

zivnevo commented on src/baseline_rule.py:
```diff
@@ -88,6 +107,28 @@ def matches_target(self, labels):
 
+    def matches_source_namespace(self, labels):
+        """
+        Check whether the given set of labels match the rule source's namespace
```
```suggestion
        Check whether the given set of labels match the rule source's namespace
        NOTE: if no labels are provided, this is interpreted as "no namespace context", and the rule matches
```
Also add this comment in `matches_target_namespace`

zivnevo reviewed (changes requested):
Looks good - only documentation-related comments
